src/tqec/plaquette/compilation/passes/transformer/transformer.py

```diff
@@ -0,0 +1,132 @@
+from dataclasses import dataclass
+from typing import Callable, Sequence
+
+import stim
+from typing_extensions import override
+
+from tqec.circuit.moment import Moment
+from tqec.circuit.schedule.circuit import ScheduledCircuit
+from tqec.plaquette.compilation.passes.base import CompilationPass
+from tqec.plaquette.compilation.passes.transformer.schedule import ScheduleFunction
+from tqec.plaquette.compilation.passes.transformer.simplifiers import (
+    InstructionSimplifier,
+    NoInstructionSimplification,
+)
+
+
+@dataclass
+class InstructionCreator:
+    """Create an instruction from targets and arguments."""
+
+    name: str
+    targets: Callable[[list[stim.GateTarget]], list[stim.GateTarget]] = (
+        lambda trgts: trgts
+    )
+    arguments: Callable[[list[float]], list[float]] = lambda args: args
+
+    def __call__(
+        self, targets: list[stim.GateTarget], arguments: list[float]
+    ) -> stim.CircuitInstruction:
+        return stim.CircuitInstruction(
+            self.name, self.targets(targets), self.arguments(arguments)
+        )
+
+
+@dataclass
+class ScheduledCircuitTransformation:
+    """Describes an instruction transformation.
+
+    This class describes how a given instruction should be transformed into
+    potentially several instructions.
+
+    Attributes:
+        source_name: name of the instruction that is transformed by ``self``.
+        transformation: a mapping from a schedule description to a list of
+            instruction creators. The instructions created by the instruction
+            creators will be inserted at the schedule computed by the schedule
+            description.
+        instruction_simplifier: a simplifier applied before trying to create a
+            :class:`~tqec.circuit.moment.Moment` instance with the instructions
+            resulting from the application of ``self``.
+    """
+
+    source_name: str
+    transformation: dict[ScheduleFunction, list[InstructionCreator]]
+    instruction_simplifier: InstructionSimplifier = NoInstructionSimplification()
+
+    def apply(self, circuit: ScheduledCircuit) -> ScheduledCircuit:
+        # moment_instructions: schedule_index -> instruction list.
+        moment_instructions: dict[int, list[stim.CircuitInstruction]] = {}
+        for schedule, moment in circuit.scheduled_moments:
+            for instruction in moment.instructions:
+                # if the transformation represented by self does not apply to
+                # the current instruction, just add it unmodified.
+                if instruction.name != self.source_name:
```

Comment:
```suggestion
                if instruction.name not in stim.gate_data(self.source_name).aliases:
```

This eliminates the need to specify transformation rules both for `R` and `RZ`.